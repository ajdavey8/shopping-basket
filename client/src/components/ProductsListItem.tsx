import React from 'react'
import styled from 'styled-components';
import { AddItem, RemoveItem } from '../types';

import { Button, ItemContainer } from './styled-components'

const TextWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: 16px;
`

const ButtonsWrapper = styled.div`
  display: flex;
  flex-direction: column;
`

interface ProductsListItem {
  id: number
  name: string
  description: string
  price: number
  image: string
  addItem: (item: AddItem) => void
  removeItem: (item: RemoveItem) => void
}

const ProductsListItem = ({ id, name, description, price, image, addItem, removeItem }: ProductsListItem) => {
  const onAddItem = () => {
    const item = {
      id,
      image,
      name,
      price
    }
    addItem(item)
  }

  const onRemoveItem = () => {
    const item = {
      id,
      price
    }
    removeItem(item)
  }

  return (
    <ItemContainer key={id}>
      <img src={image} alt={`${name} image`} />
      <TextWrapper>
        <h2>{name}</h2>
        <p>{description}</p>
      </TextWrapper>
      <ButtonsWrapper>
        <Button onClick={onAddItem}>Add</Button>
        <Button onClick={onRemoveItem}>Remove</Button>
      </ButtonsWrapper>
    </ItemContainer>
  )
}

export default ProductsListItem