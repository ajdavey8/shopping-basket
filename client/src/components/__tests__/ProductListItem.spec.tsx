import React from 'react';
import { render, fireEvent } from '@testing-library/react'
import ProductsListItem from '../ProductsListItem';

const addItem = jest.fn()
const removeItem = jest.fn()

const setup = () => {
  const utils = render(
    <ProductsListItem
      name='product'
      description='describe'
      id={1}
      price={2}
      image='imageURL'
      addItem={addItem}
      removeItem={removeItem}
    />
  )

  return {
    ...utils
  }
}

describe('<ProductsListItem>', () => {
  it('renders with the passed props', () => {
    const { asFragment } = setup()
    expect(asFragment).toMatchSnapshot();
  })

  it('calls the onClick function to add item', () => {
    const { getByText } = setup()

    fireEvent.click(getByText('Add'))
    expect(addItem).toHaveBeenCalledWith({ id: 1, price: 2, name: 'product', image: 'imageURL' })
  })

  it('calls the onClick function to remove item', () => {
    const { getByText } = setup()

    fireEvent.click(getByText('Remove'))
    expect(removeItem).toHaveBeenCalledWith({ id: 1, price: 2 })
  })
})