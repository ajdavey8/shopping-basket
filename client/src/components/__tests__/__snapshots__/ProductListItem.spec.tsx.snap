// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsListItem> renders with the passed props 1`] = `[Function]`;
